prometheus/example_timer_batch_test.go

```diff
@@ -0,0 +1,48 @@
+// Copyright 2014 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus_test
+
+import (
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	// For periodic or batch processes that run much less frequently than
+	// the Prometheus scrape, it makes sense to use a Gauge to observe the
+	// duration.
+	// See https://prometheus.io/docs/practices/instrumentation/#batch-jobs
+	batchDuration = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "example_batch_duration_seconds",
+		Help: "Duration of the last batch run.",
+	})
+)
+
+func performBatch() {
+	// The Set method of the Gauge is used to observe the duration.
+	timer := prometheus.NewTimer(prometheus.ObserverFunc(batchDuration.Set))
+	defer timer.ObserveDuration()
```

Comment:
> Handling those two things in one Timer helper will be a nice gain, and would totally justify to have a dedicated Gauge-Timer (perhaps call it a BatchTimer or something).

The other common case for the current time is recording when you last did something (e.g. processed a record from a queue) to detect when processing has broken. I don't think I'd mix that in with batch jobs, I can see it getting messy when a small bit of extra client code would do the same task in a clearer way.

> So would you suggest two separate metrics for success and failure timestamp?

I'd have a completed, and a success.